<a href="https://colab.research.google.com/github/DurgaBhavana/5731Submissions/blob/master/In_class_exercise_08.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The eighth in-class-exercise (20 points in total, 10/29/2020)**

The data for this exercise is from the dataset you created from assignment three. Please perform answer the following questions based on your data:

## (1) (10 points) Write a python program to extract the sentiment related terms from the corpus. You may use python package such as polyglot or external lexicon resources in the question. Rank the sentiment related terms by frequency.

In [40]:
import pandas as pd
df = pd.read_csv('/content/Abstracts for Sentiment Analysis.csv')
df

document_id                                         clean_text sentiment
0     Abstract 1  describe a method for statistical modeling bas...  Positive
1     Abstract 2  Scaling conditional random fields for natural ...   Neutral
2     Abstract 3  The paper addresses the issue of cooperation b...  Positive
3     Abstract 4  In most natural language processing applicatio...  Positive
4     Abstract 5  We propose a unified neural network architectu...  Positive
..           ...                                                ...       ...
95   Abstract 96  This paper presents a workbench built by Pribe...  Positive
96   Abstract 97  Abstract—Natural Language Processing (NLP) is ...  Positive
97   Abstract 98  ABSTRACT: After twenty years of disfavor, a te...  Negative
98   Abstract 99  Text statistics are frequently used in stylome...   Neutral
99  Abstract 100  We summarize our experience using FrameNet in ...  Negative

[100 rows x 3 columns]

In [41]:
# Pre-processing
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')
import re
from textblob import TextBlob
from textblob import Word
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [42]:
# Special characters removal
df['After noise removal'] = df['clean_text'].apply(lambda x: ''.join(re.sub(r"[^a-zA-Z0-9]+", ' ', charctr) for charctr in x ))
# Punctuation removal
df['Punctuation removal'] = df['After noise removal'].str.replace('[^\w\s]','')
# Remove numbers
df['Remove numbers'] = df['Punctuation removal'].str.replace('\d+', '')
# Stopwords removal
stop_word = stopwords.words('english')
df['Stopwords removal'] = df['Remove numbers'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_word))
# Lower Casing
df['Lower casing'] = df['Stopwords removal'].apply(lambda x: " ".join(x.lower() for x in x.split()))
# Lemmatization
df['Lemmatization'] = df['Lower casing'].apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()]))
df

document_id  ...                                      Lemmatization
0     Abstract 1  ...  describe method statistical modeling based max...
1     Abstract 2  ...  scaling conditional random field natural langu...
2     Abstract 3  ...  the paper address issue cooperation linguistic...
3     Abstract 4  ...  in natural language processing application des...
4     Abstract 5  ...  we propose unified neural network architecture...
..           ...  ...                                                ...
95   Abstract 96  ...  this paper present workbench built priberam in...
96   Abstract 97  ...  abstract natural language processing nlp effec...
97   Abstract 98  ...  abstract after twenty year disfavor technology...
98   Abstract 99  ...  text statistic frequently used stylometry cryp...
99  Abstract 100  ...  we summarize experience using framenet two rat...

[100 rows x 9 columns]

In [73]:
from textblob import TextBlob
from collections import Counter

In [85]:
sentiment_text = []
for i in df['Lemmatization']:
  for term in i.split(' '):
    if TextBlob(term).sentiment.polarity != 0:
      sentiment_text.append(term)
print(sentiment_text)
word_count = Counter(sentiment_text)
word_df = pd.DataFrame(list(word_count.most_common()), columns=['Terms', 'Freq'])
word_df.index = list(range(1, len(word_count.most_common())+1))
print(word_df)

['natural', 'random', 'natural', 'natural', 'general', 'particular', 'natural', 'base', 'natural', 'more', 'natural', 'natural', 'subject', 'natural', 'broad', 'narrow', 'broad', 'natural', 'natural', 'natural', 'natural', 'unprocessed', 'natural', 'natural', 'natural', 'able', 'natural', 'natural', 'natural', 'natural', 'linguistic', 'directly', 'appropriate', 'natural', 'better', 'vague', 'precise', 'natural', 'natural', 'natural', 'lyric', 'significant', 'lyric', 'lyric', 'important', 'cultural', 'linguistic', 'direct', 'detailed', 'natural', 'high', 'interesting', 'new', 'natural', 'natural', 'naturally', 'natural', 'modern', 'limited', 'behind', 'limited', 'intelligent', 'natural', 'military', 'natural', 'developed', 'base', 'high', 'significant', 'far', 'particular', 'much', 'particular', 'natural', 'main', 'natural', 'naturally', 'linguistic', 'natural', 'kind', 'large', 'natural', 'contemporary', 'natural', 'natural', 'new', 'relevant', 'sound', 'developed', 'natural', 'natural

## (2) (10 points) Compare the performance of the following tools in sentiment identification: TextBlob (https://textblob.readthedocs.io/en/dev/), VADER (https://github.com/cjhutto/vaderSentiment), TFIDF-based Support Vector Machine (SVM) (Split your data into training and testing data). Take your own annotation as the standard answers. 

Reference code: https://towardsdatascience.com/fine-grained-sentiment-analysis-in-python-part-1-2697bb111ed4

In [98]:
def get_sentiments_counts(sentiments):
  Negative_value = Positive_value = Neutral_value = 0
  for polarity in sentiments:
    if polarity < 0:
      Negative_value = Negative_value + 1
    elif polarity > 0:
      Positive_value = Positive_value + 1
    elif polarity == 0:
      Neutral_value = Neutral_value + 1
  return Negative_value, Positive_value, Neutral_value

# TEXTBLOB

In [100]:
nltk.download('vader_lexicon')
from textblob import TextBlob
sentiments = []
for i in df['Lemmatization']:
  sentiments.append(TextBlob(i).sentiment.polarity)
Negative_value, Positive_value, Neutral_value = get_sentiments_counts(sentiments)
print('Number of Positives - {0}'.format(Positive_value))
print('Number of Negatives - {0}'.format(Negative_value))
print('Number of Neutrals - {0}'.format(Neutral_value))

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!
Number of Positives - 75
Number of Negatives - 10
Number of Neutrals - 15


# VADER

In [101]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
vader_analyzer = SentimentIntensityAnalyzer()
sentiments = []
for i in df['Lemmatization']:
  sentiments.append(vader_analyzer.polarity_scores(i)['compound'])
Negative_value, Positive_value, Neutral_value = get_sentiments_counts(sentiments)
print('Number of Positives - {0}'.format(Positive_value))
print('Number of Negatives - {0}'.format(Negative_value))
print('Number of Neutrals - {0}'.format(Neutral_value))

Number of Positives - 79
Number of Negatives - 8
Number of Neutrals - 13


/usr/local/lib/python3.6/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


# SVM

In [102]:
from sklearn.model_selection import train_test_split

In [103]:
def split_data(x_values, y_values):
  return train_test_split(x_values, y_values, test_size=0.33)

In [104]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [105]:
def get_vector():
  return TfidfVectorizer(min_df = 5, max_df = 0.8, sublinear_tf = True, use_idf = True)

In [106]:
TfidfVectorizer = get_vector()
training_x, testing_x, training_y, testing_y = split_data(df['Lemmatization'], df['sentiment'])
train_vectors = TfidfVectorizer.fit_transform(training_x)
test_vectors = TfidfVectorizer.transform(testing_x)

In [107]:
from sklearn.metrics import classification_report
from sklearn import svm

In [108]:
def train_predict_model(train_vectors, test_vectors, training_y):
  model = svm.SVC(kernel='linear')
  model.fit(train_vectors, training_y)
  return model.predict(test_vectors)

In [109]:
svm_report = classification_report(testing_y, train_predict_model(train_vectors, test_vectors, training_y), output_dict=True)

In [117]:
print('positive Report - ', svm_report['Positive'])
print('Negative Report - ', svm_report['Negative'])
print("Neutral Report - ", svm_report['Neutral'])

positive Report -  {'precision': 0.6451612903225806, 'recall': 0.9523809523809523, 'f1-score': 0.7692307692307692, 'support': 21}
Negative Report -  {'precision': 1.0, 'recall': 0.14285714285714285, 'f1-score': 0.25, 'support': 7}
Neutral Report -  {'precision': 0.0, 'recall': 0.0, 'f1-score': 0.0, 'support': 5}


In [114]:
def get_actual_sentiment_counts(sentiments):
  Negative_value = Positive_value = Neutral_value = 0
  for polarity in sentiments:
    if polarity == 'Negative':
      Negative_value = Negative_value + 1
    elif polarity == 'Positive':
      Positive_value = Positive_value + 1
    elif polarity == 'Neutral':
      Neutral_value = Neutral_value + 1
  return Negative_value, Positive_value, Neutral_value

In [115]:
Actual_Negative_Value, Actual_Positive_Value, Actual_Neutral_Value = get_actual_sentiment_counts(df['sentiment'])

In [116]:
print('Number of Actual Positives - {0}'.format(Actual_Positive_Value))
print('Number of Actual Negatives - {0}'.format(Actual_Negative_Value))
print('Number of Actual Neutrals - {0}'.format(Actual_Neutral_Value))

Number of Actual Positives - 59
Number of Actual Negatives - 25
Number of Actual Neutrals - 16


**Results:**

Actual - 

Number of Actual Positives - 59

Number of Actual Negatives - 25

Number of Actual Neutrals - 16


TextBlob - 

Number of Positives - 75

Number of Negatives - 10

Number of Neutrals - 15


Vader -

Number of Positives - 79

Number of Negatives - 8

Number of Neutrals - 13

SVM - 

positive Report -  {'precision': 0.6451612903225806, 'recall': 0.9523809523809523, 'f1-score': 0.7692307692307692, 'support': 21}

Negative Report -  {'precision': 1.0, 'recall': 0.14285714285714285, 'f1-score': 0.25, 'support': 7}

Neutral Report -  {'precision': 0.0, 'recall': 0.0, 'f1-score': 0.0, 'support': 5}


From above results - 

On whole, we can say that SVM has better accuracy of polarities for the abstract texts.
To conclude comparisons, we can say that SVM is better compared to TextBlob and Vader and among TextBlob and Vader, Textblob is better
